pkg/arvo/app/chat-push-hook.hoon

```diff
@@ -0,0 +1,255 @@
+/-  *permission-store, *invite-store, hook=chat-push-hook
+/+  default-agent, verb, dbug, store=chat-store, lib=chat-hooks
+|%
++$  card  card:agent:gall
++$  versioned-state
+  $%  [%0 state-0]
+  ==
++$  state-0
+  $:  sharing=(set path)
+      allow-history=(map path ?)
+  ==
+--
+::
+=|  [%0 state-0]
+=*  state  -
+::
+%-  agent:dbug
+%+  verb  |
+::
+^-  agent:gall
+|_  bol=bowl:gall
++*  this  .
+    def  ~(. (default-agent this %|) bol)
+::
+++  on-init
+  ^-  (quip card _this)
+  :_  this
+  [%pass /permissions %agent [our.bol %permission-store] %watch /updates]~
+::
+++  on-save  !>(state)
+++  on-load
+  |=  vax=vase
+  =/  old  !<(versioned-state vax)
+  ?-  -.old
+    %0  [~ this(state old)]
+  ==
+::
+++  on-poke
+  |=  [=mark =vase]
+  ^-  (quip card _this)
+  |^
+  =^  cards  state
+    ?+  mark  (on-poke:def mark vase)
+        %json  (chat-action (action:dejs:store !<(json vase)))
+        %chat-action  (chat-action !<(action:store vase))
+        %chat-push-hook-action  (chat-push-hook-action !<(action:hook vase))
+    ==
+  [cards this]
+  ::
+  ++  chat-action
+    |=  act=action:store
+    ^-  (quip card _state)
+    ::  XX why do we use ?> here but ?. ~ below?
+    ?>  ?=(%message -.act)
+    :_  state
+    ?.  (~(has in sharing) path.act)  ~
+    ?.  (is-permitted:lib bol src.bol path.act)  ~
+    ::  XX do we still need this part?
+    =*  letter  letter.envelope.act
+    =?  letter
+        ?&  (team:title our.bol src.bol)
+            ?=(%code -.letter)
+            ?=(~ output.letter)
+        ==
+      =/  =hoon  (ream expression.letter)
+      letter(output (eval:store bol hoon))
+    =:  author.envelope.act  src.bol
+        when.envelope.act  now.bol
+    ==
+    [%pass / %agent [our.bol %chat-store] %poke %chat-action !>(act)]~
+   
+  ::
+  ++  chat-push-hook-action
+    |=  act=action:hook
+    ^-  (quip card _state)
+    ?-  -.act
+        %add
+      ?>  (team:title our.bol src.bol)
+      =/  chat-path  [%mailbox path.act]
+      =/  chat-wire  [%store path.act]
+      ?:  (~(has in sharing) path.act)  [~ state]
+      =:  sharing  (~(put in sharing) path.act)
+          allow-history  (~(put by allow-history) path.act allow-history.act)
+      ==
+      :_  state
+      [%pass chat-wire %agent [our.bol %chat-store] %watch chat-path]~
+    ::
+        %remove
+      ?.  !(team:title our.bol src.bol)
+        [~ state]
+      ?.  (~(has in sharing) path.act)
+        ~&  [dap.bol %already-not-sharing path.act]
+        [~ state]
+      =.  sharing  (~(del in sharing) path.act)
+      :_  state
+      :~  [%give %kick [%mailbox path.act]~ ~]
+          [%pass [%mailbox path.act] %agent [our.bol %chat-store] %leave ~]
+          ::  XX is it worth kicking backlog subscriptions here just to be
```

Comment:
Yes, good idea. You’d have to do a similar iteration over the current ones to what is written in the pull-hook though